src/include/storage/sql_table.h

```diff
@@ -106,191 +110,49 @@ class SqlTable {
    * @param schema the initial Schema of this SqlTable
    * @param oid unique identifier for this SqlTable
    */
-  SqlTable(BlockStore *const store, const catalog::Schema &schema, const catalog::table_oid_t oid)
-      : block_store_(store), oid_(oid), schema_version_(0) {
-    const auto layout_and_map = StorageUtil::BlockLayoutFromSchema(schema);
-
-    DataTableVersion first_table = {new DataTable(block_store_, layout_and_map.first, layout_version_t(0)),
-                                    layout_and_map.first, layout_and_map.second};
-    tables_.emplace_back(first_table);
-  }
+  SqlTable(BlockStore *store, const catalog::Schema &schema, catalog::table_oid_t oid);
 
   /**
    * Destructs a SqlTable, frees all its members.
    */
-  ~SqlTable() {
-    for (auto &t : tables_) delete t.data_table;
-  }
+  ~SqlTable();
+
+  /**
+   * Adds the new schema to set of active data tables.  This functions should
+   * only be called upon commit because there is no mechanism to rollback an
```

Comment:
Why is it hard to roll back? I think deleting it from the map should be sufficient. Right? Also, I think whether a transaction updating the schema needs to abort depends on other concurrent transactions and the rules we set on who needs abort. I don't remember but did we agree that a transaction updating the schema always commits?

Reply:
Ordering of `UpdateSchema` with `commit`:  Right now I see `UpdateSchema` being called along the critical path of `commit`.  While this could be performance problem, it avoids needing to create special logic for handling an UndoRecord for the catalog (i.e. deleting something from a map).

Inserting a tuple after a schema change:  Talking with Matt, we're not sure whether that would actually be a valid transaction.  Specifically, if the schema version for a transaction is fixed based off its start time, a transaction that modifies tables would be unable to semantically see the new schema versions until it commits.  If we want to support that I think it's a larger group discussion.

Rolling back is hard(er) at the moment because it would mean adding more code to the handling of UndoRecords.  Specifically, if a schema change aborted (requiring a roll back) we would need to detect whether or not it is aborting before or after having staged the new version (which is not covered semantically by anything at the moment) and handle accordingly.  With regard to whether a schema always commit, we agreed that "safe" changes should always commit (i.e. no constraints or type changes) and punted the ball on "unsafe" changes until we had implementation proposals (possible need for partial serialization).

Reply:
Regarding the need for a concurrency test, I agree with your assessment that we need a concurrent test for SQL table.  However, I think that the test we had was only testing multiple schema changes where as we probably want to simply port the data table concurrent test up to SQL table and inject a few schema changes in to the stream to make sure it properly handles transactions in-flight at the same time as the schema change.